lib/config/rule-validator.js

```diff
@@ -0,0 +1,155 @@
+ * @param {string} ruleId The ID of the rule to find.
+ * @param {Object} config The config to search in.
+ * @returns {{create: Function, schema: (Array|null)}} THe rule object.
+ */
+function findRuleDefinition(ruleId, config) {
+    const ruleIdParts = ruleId.split("/");
+    let pluginName, ruleName;
+
+    // built-in rule
+    if (ruleIdParts.length === 1) {
+        pluginName = "@";
+        ruleName = ruleIdParts[0];
+    } else {
+        ruleName = ruleIdParts.pop();
+        pluginName = ruleIdParts.join("/");
+    }
+
+    if (!config.plugins || !config.plugins[pluginName]) {
+        throw new TypeError(`Key "rules": Key "${ruleId}": Could not find plugin "${pluginName}".`);
+    }
+
+    if (!config.plugins[pluginName].rules || !config.plugins[pluginName].rules[ruleName]) {
+        throw new TypeError(`Key "rules": Key "${ruleId}": Could not find "${ruleName}" in plugin "${pluginName}".`);
+    }
+
+    return config.plugins[pluginName].rules[ruleName];
+
+}
+
+/**
+ * Gets a complete options schema for a rule.
+ * @param {{create: Function, schema: (Array|null)}} rule A new-style rule object
+ * @returns {Object} JSON Schema for the rule's options.
+ */
+function getRuleOptionsSchema(rule) {
+
+    if (!rule) {
+        return null;
+    }
+
+    const schema = rule.schema || rule.meta && rule.meta.schema;
+
+    // Given a tuple of schemas, insert warning level at the beginning
+    if (Array.isArray(schema)) {
+        if (schema.length) {
+            return {
+                type: "array",
+                items: schema,
+                minItems: 0,
+                maxItems: schema.length
+            };
+        }
+        return {
+            type: "array",
+            minItems: 0,
+            maxItems: 0
+        };
+
+    }
+
+    // Given a full schema, leave it alone
+    return schema || null;
+}
+
+//-----------------------------------------------------------------------------
+// Exports
+//-----------------------------------------------------------------------------
+
+/**
+ * Implements validation functionality for the rules portion of a config.
+ */
+class RuleValidator {
+
+    /**
+     * Creates a new instance.
+     */
+    constructor() {
+
+        /**
+         * A collection of compiled validators for rules that have already
+         * been validated.
+         * @type {WeakMap}
+         * @property validators
+         */
+        this.validators = new WeakMap();
+    }
+
+    /**
+     * Validates all of the rule configurations in a config against each
+     * rule's schema.
+     * @param {Object} config The full config to validate. This object must
+     *      contain both the rules section and the plugins section.
+     * @returns {void}
+     * @throws {Error} If a rule's configuration does not match its schema.
+     */
+    validate(config) {
+
+        if (!config.rules) {
+            return;
+        }
+
+        for (const ruleId of Object.keys(config.rules)) {
```

Comment:
We are currently not doing any validations for rules that are set to `0`/`"off"` severity.

This configuration is valid with the actual config system:

```js
rules: {
        "dot-location": [0, "foo"], // invalid option
        "foo": 0, // rule doesn't exist
        "foo/foo": 0, // plugin doesn't exist
    }
```

Reply:
Got it. Thanks.